### ISSUE TITLE
ErrorExtension support for InputValueValidator

### ISSUE THREAD

**Issue by mjonuschat** [async-graphql/async-graphql#385]
## Description of the feature

Support returning an Error with  ErrorExtensions from (custom) validators to provide more (machine parseable) information back to the client. Parsing the stringified error message to gather information like the field or validation error type is brittle and error prone. Also return machine parsable information for the argument and path by default.

### Current response:
```json
{
  "data": null,
  "errors": [
    {
      "message": "Invalid value for argument \"request.email\", invalid email format",
      "locations": [
        {
          "line": 2,
          "column": 23
        }
      ]
    }
  ]
}
```

### Desired response
```json
{
  "data": null,
  "errors": [
    {
      "message": "Invalid value for argument \"request.email\", invalid email format",
      "code": "INVALID_EMAIL_FORMAT",
      "argument": "request.email",
      "path": ["request", "email"],
      "locations": [
        {
          "line": 2,
          "column": 23
        }
      ]
    }
  ]
}
```

## Code example (if possible)
```rust
impl InputValueValidator for CustomValidator {
    fn is_valid(&self, value: &Value) -> Result<(), Error> {
        return Err(Error::new("Some failure reason").extend_with(|_, e| {
            e.set("code", "OUT_OF_RANGE")
        }));
    }
}
```

**Comment by sunli829** [comment]
I need to think about how to add this feature elegantly.

**Comment by cptrodgers** [comment]
@mjonuschat  I think this issue is relating to https://github.com/async-graphql/async-graphql/pull/564. I guess it will be released on 2.9.8.
It similar to your proposal with some minor different:

Code example
```rust
impl InputValueValidator for CustomValidator {
    fn is_valid_with_extentions(&self, value: &Value) -> Result<(), Error> {
        return Err(Error::new("Some failure reason").extend_with(|_, e| {
            e.set("code", "OUT_OF_RANGE")
        }));
    }
}
```

Output expected:
```json
{
  "data": null,
  "errors": [
    {
      "message": "Invalid value for argument \"request.email\", invalid email format",
      
      "argument": "request.email",
      "path": ["request", "email"],
      "locations": [
        {
          "line": 2,
          "column": 23
        }
      ],
      extensions: {"code": "INVALID_EMAIL_FORMAT"}
    }
  ]
}
```

**Comment by mjonuschat** [comment]
@cptrodgers This looks great and is very close to what I was envisioning with this feature suggestion. I'll close this in favor of #564
